lib/utils/note-utils.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * Matches the title and excerpt in a note's content
+ *
+ * Both the title and the excerpt are determined as
+ * content starting with something that isn't
+ * whitespace and leads up to a newline or line end
+ *
+ * @type {RegExp} matches a title and excerpt in note content
+ */
+const noteTitleAndPreviewRegExp = /\s*(\S[^\n]*)\s*(\S[^\n]*)?/;
+
+/**
+ * Matches the title and excerpt in a note's content skipping opening # from title
+ *
+ * Both the title and the excerpt are determined as
+ * content starting with something that isn't
+ * whitespace and leads up to a newline or line end
+ * @type {RegExp} matches a title and excerpt in note content skipping opening #'s from title and preview
+ */
+const noteTitleAndPreviewMdRegExp = /[#]{0,6}\s+?(\S[^\n]*)[\s]*[#]{0,6}\s*?(\S[^\n]*)?/;
```

Comment:
this comment is more for my own note-taking as I review the RegExp

```js
const noteTitleAndPreviewMdRegExp = new RegExp(
	'[#]{0,6}\s+?' + // start matching when we have a header with space after it (ungreedy)
	'(' + // match group 1 - title
		'\S[^\n]*' + // title is some non-whitespace character and then everything until the end of the line
	')' +
	'[\s]*' + // followed by some amount of whitespace including the newline
	'[#]{0,6}\s*?' + // the next bit might be prefixed by a header, and if it is we won't use it for the excerpt
	'(' + // match group 2 - preview
		'\S[^\n]*' + // preview is the first non-whitespace character and then everything until the end of the line
	')?' // there may or may not be an excerpt
);
```

actually now I'm wondering if this is really what we want. it may also match on `content: '$# not a title'`

**questions**
 - is there a reason we have the `[]` where they aren't required, like around the `#` and `\s`?
 - should we be processing the first line in two stages for Markdown notes? It seems like this RegExp pattern isn't handling Markdown matches properly when they aren't prefixed by the hash.

Markdown Title:
 - Skip all whitespace prefixing content
 - Grab first line
 - If prefixed by hashes that would indicate a header then strip them

Markdown Preview:
 - Skip whitespace after title
 - Grab until next newline
 - If prefixed by hashes that would indicate a header then strip them

Reply:
```js
const match = /\s*(\S[^\n]*)\s*(\S[^\n]*)?/.exec( content || '' );
if ( ! match ) {
	return defaults;
}

const [ /* full match */, first, second ] = match;

const title = isMarkdown
	? first.replace( /^#{1,6}\s+/, '' )
	: first;

const preview = isMarkdown
	? second.replace( /^#{1,6}\s+/, '' )
	: second;

return {
	title: title.slice( 0, 200 ) || default.title,
	preview: preview || default.preview
}
```

thoughts?

Reply:
`[]` are not required. You're right that this one won't work. Here is form that will work: `/(?>#{0,6}\s+)?(\S[^\n]*)\s*(?>#{0,6}\s+)?(\S[^\n]*)?/`. Please take a look on that @dmsnell 

I think that doing one regex is better than making 3 but I'm not an expert.

Reply:
Sadly I don't think that `(?>…)` exists in Javascript.

One vs. three RegExps isn't usually clear; it's not always better to have fewer and it depends on the characteristics of the expression _and_ the document.

In this case yours would probably be fine and faster though I just end up questioning the complexity of the source code. I don't know that this is a performance problem and I find the logic simpler when operating on the line-by-line basis - I almost also wrote `content.split('\n').slice(0,2)` up there too!

Actually I'm pretty sure that performance shouldn't be an issue here. After the initial note list generation it should cache the previews for the notes and the only one being recalculated should be the one that's in the editor changing or coming in from the server in updates.

We can replace the "atomic groups" in this new RegExp with non-capture groups which are similar… `(?:#{0,6}\s+)…` They might get the job done.

Reply:
You're right that with caching any approach should be enough fast for this simple task. But those regex's are not so complicated so, for now, I'll live it as it is. And I'll use non-capture groups instead of atomic groups which don't exist in js.